src/backend/tcop/zenith_wal_redo.c

```diff
@@ -683,6 +683,9 @@ GetPage(StringInfo input_message)
 	page = BufferGetPage(buf);
 	/* single thread, so don't bother locking the page */
 
+	// Calculate checksums on the newly materialized page if needed.
+	PageSetChecksumInplace(page, blknum);
```

Comment:
If we want to be pedantic we should validate checksum before running redo (in PushPage)
also, Log record CRC should be validated as well, currently it's only done in XLogReadRecord but not DecodeXLogRecord (used directly in ApplyRecord)

Reply:
> If we want to be pedantic we should validate checksum before running redo (in PushPage)

IIUC, checksum can be not up-to-date in the full-page-image inside WAL record, as Postgres makes the final update on them only on eviction. That's the reason actually, why I stuck with these PRs. On pageserver we take FPIs from WAL records and store them inside timeline data as 'pages', not original WAL records. So next time we pass these pages back to the Postgres, it complains that they have a wrong CRC. Yet, if I force WAL redo for them, everything is fine, as we stamp a proper CRC on them in the GetPage()

> also, Log record CRC should be validated as well, currently it's only done in XLogReadRecord but not DecodeXLogRecord (used directly in ApplyRecord)

Good catch, thanks! I'll have a look on this part too